utils/src/main/java/software/amazon/awssdk/utils/AttributeMap.java

```diff
@@ -199,17 +266,291 @@ public <T> Builder put(Key<T> key, T value) {
+                attributesToCopy.forEach((k, v) -> {
+                    Value<?> newValue = v.copy();
+                    valueRemapping.put(v, newValue);
+                    attributes.put(k, newValue);
+                });
+                dependencyGraph = dependencyGraph.copy(valueRemapping::get);
+                copyOnUpdate = false;
+            }
+        }
+
         @Override
         public AttributeMap build() {
-            return new AttributeMap(configuration);
+            AttributeMap result = new AttributeMap(this);
+            copyOnUpdate = true;
+            return result;
+        }
+    }
+
+    /**
+     * A value that is evaluated lazily. See {@link Builder#putLazy(Key, LazyValue)}.
+     */
+    @FunctionalInterface
+    public interface LazyValue<T> {
+        T get(LazyValueSource source);
+    }
+
+    /**
+     * A source for other values, provided to a {@link LazyValue} when the value is resolved.
+     */
+    @FunctionalInterface
+    public interface LazyValueSource {
+        <T> T get(Key<T> sourceKey);
+    }
+
+    /**
+     * Tracks which values "depend on" other values, so that when we update one value, when can clear the cache of any other
+     * values that were derived from the value that was updated.
+     */
+    private static final class DependencyGraph {
+        /**
+         * Inverted adjacency list of dependencies between derived keys. Mapping from a key to what depends on that key.
+         */
+        private final Map<Key<?>, Set<Value<?>>> dependents;
+
+        private DependencyGraph() {
+            this.dependents = new HashMap<>();
+        }
+
+        private DependencyGraph(Map<Key<?>, Set<Value<?>>> dependents,
+                                Function<Value<?>, Value<?>> valueRemapping) {
+            this.dependents = new HashMap<>(dependents.size());
+            dependents.forEach((key, values) -> {
+                Set<Value<?>> newValues = new HashSet<>(values.size());
+                this.dependents.put(key, newValues);
+                values.forEach(v -> newValues.add(valueRemapping.apply(v)));
+            });
+        }
+
+        private void addDependency(Value<?> consumer, Key<?> dependency) {
+            dependents.computeIfAbsent(dependency, k -> new HashSet<>()).add(consumer);
+        }
+
+        private Set<Value<?>> dependents(Key<?> key) {
+            return dependents.getOrDefault(key, emptySet());
+        }
+
+        private void invalidateConsumerCaches(Key<?> key) {
+            Set<Value<?>> dependents = dependents(key);
+            if (dependents.isEmpty()) {
+                return;
+            }
+
+            Queue<Value<?>> unloadQueue = new ArrayDeque<>(dependents);
+            dependents.clear();
+            while (!unloadQueue.isEmpty()) {
+                Value<?> toUnload = unloadQueue.poll();
+
+                Set<Value<?>> toUnloadDependents = dependents(toUnload.key());
```

Comment:
Is it worth optimizing here by avoiding unloading values more than once?

Reply:
I don't think so. Unloading the value is pretty much free, since we're just clearing the cache and not actually refreshing the value until it's requested again.